Timesync improvements - part 3

fessehaeve opened the pull request:
Fixes for issues form https://github.com/project-chip/connectedhomeip/issues/27575 :

- improved time zone name changing check
- improved time zone name copying buffer size comparison
- removed EventFlag enum and its usage - https://github.com/project-chip/connectedhomeip/pull/26082/files#r1242657708 
- changed mechanism by which time zone and dst tables are maintained - https://github.com/project-chip/connectedhomeip/pull/26082#discussion_r1242628037, https://github.com/project-chip/connectedhomeip/pull/26082/files#r1242655854
- event generation is now part of maintaining time zone and dst tables. In some cases, event generation is handled by the caller of the functions that are updating the tables - https://github.com/project-chip/connectedhomeip/pull/26082/files#r1242683543
- updated test that is checking validity of dst offset items
- store one less granularity in SetUTCTime command - https://github.com/project-chip/connectedhomeip/pull/26082/files#r1242671601, https://github.com/project-chip/connectedhomeip/pull/26082/files#r1242676471
- use the correct error types - https://github.com/project-chip/connectedhomeip/pull/26082#discussion_r1242631916
- setting empty time zone list should result in ConstraintError
- simplify redundant code for loading TZ and DST
- https://github.com/project-chip/connectedhomeip/pull/26082#discussion_r1241227274
- make TimeSource attribute use AAI
- make it possible to initialise time sync server with any persistent storage implementation - rather than getting the default from Server


#### Testing 

This is tested by cert tests and unit tests

fessehaeve committed:
setting empty time zone list should result in ConstraintError

fessehaeve committed:
- use the correct error types
- simplify redundant code with loading TZ and DST

fessehaeve committed:
store one less granularity in SetUTCTime command

fessehaeve committed:
- removed EventFlag enum and its usage
- changed mechanism by which time zone and dst tables are maintained
- event generation is now part of maintaining time zone and dst tables. In
 some cases, event generation is handled by the caller of the functions
 that are updating the tables
- updated test that is checking validity of dst offset items

fessehaeve committed:
- improved time zone name changing check
- improved time zone name copying buffer size comparison

fessehaeve committed:
restyle

fessehaeve commented:
FYI - I am still figuring out a way to add unit tests to this.

fessehaeve committed:
make it possible to init time sync server with any persistent storage implementation

fessehaeve committed:
stop using attribute storage for TimeSource to make the code more
suitable for unit testing

fessehaeve committed:
added read handler for TimeSource in AAI

fessehaeve commented:
> FYI - I am still figuring out a way to add unit tests to this.

The code can be tested right now with the changes I made. However, I still don't know what to test that is not covered by cert or yaml tests. Any tips would be appreciated.

cecille commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -551,15 +543,17 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
-        if (tzStore.timeZone.name.HasValue())
+        const TimeSyncDataProvider::TimeZoneStore & tzStore = GetTimeZone()[0];
+        lastTz.offset                                       = tzStore.timeZone.offset;
+        if (tzStore.timeZone.name.HasValue() && sizeof(name) >= sizeof(tzStore.name))
```
ok, so there's a resize(?) line in GetTimeZone()
    mTimeZoneObj.timeZoneList = mTimeZoneObj.timeZoneList.SubSpan(0, mTimeZoneObj.validSize);

Why not just move that re-size into where the list(s) are actually re-sized? Then the list is properly resized after the call to UpdateTimeZoneState and the getter function has no side effects. At that point, you're just directly using a class-scoped object with (I think) a class-owned buffer that won't go out of scope. And you can just directly use it as lastTz without making a copy.

Reasoning here - 1) Getters with side effects are misleading and 2) memcpy is a big bug source and makes me nervous.

If that's too much for this change, then can this at least be swapped over to use CopyCharSpanToMutableCharSpan?

cecille replied:
why would you need to re-size in multiple places if you move the resize into the update?

cecille commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -570,20 +564,17 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
         {
-            ReturnErrorOnFailure(LoadTimeZone());
-            return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_IB;
+            ExitNow(err = CHIP_IM_GLOBAL_STATUS(ConstraintError));
```
so some of these error messages have changed - do tests not complain about this? Are we not testing this? Or is this caught and changed elsewhere?

fessehaeve replied:
These errors are checked in the ember callbacks and that has been updated. So it doesn't bring change in the overall outcome.

cecille commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -647,7 +648,9 @@ CHIP_ERROR TimeSynchronizationServer::LoadTimeZone()
     InitTimeZone();
-    return mTimeSyncDataProvider.StoreTimeZone(GetTimeZone());
+    ReturnErrorOnFailure(mTimeSyncDataProvider.StoreTimeZone(GetTimeZone()));
+    emitTimeZoneStatusEvent(GetDelegate()->GetEndpoint());
```
I think the only place this function gets called right now is on Init() if the time zone fails to load. Is the intent to have that emit an event there? That seems odd. If that IS the intent, would it make more sense to move that up into the init?

This is a public method - can the app call this? Under what circumstances? If the intent is to allow this to be called by the application code, then it should probably push this though the SetTimeZone function so it can check if the current time zone is set.

fessehaeve replied:
Agree that emitting event during init is odd. Will remove it. 
My intention was that apps could decide to clear time zone for whatever reason. SetTimeZone can only be helpful if the app has the means to get a valid time zone setting without using timesync.

cecille commented on examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:
```diff
@@ -4915,10 +4915,10 @@
               "type": "TimeSourceEnum",
               "included": 1,
-              "storageOption": "RAM",
+              "storageOption": "External",
```
This is from mTimeSource now, right? So that's RAM.

fessehaeve replied:
I don't think so. RAM means it is using attribute storage. Here it is using AAI and we don't need the accessors as well.

fessehaeve committed:
- avoid using memcpy and use CopyCharSpanToMutableCharSpan instead
- do not emit TimeZoneStatus event during init
- use raw list instead of getter

andy31415 commented:
This is a 6-month old PR that has merge conflicts. Closing as stale.

bzbarsky-apple commented:
We still need to do this; it has been blocked on review from me because I could not figure out a sane way to review it, but I will get past that.

cecille reviewed (commented):
This generally looks good to me - good simplification. @andreilitvin - can you take a peek at the persistent storage stuff?

andy31415 commented:
@fessehaeve this is now a very old PR ... could you update it? Should it be maintained?

fessehaeve commented:
> @fessehaeve this is now a very old PR ... could you update it? Should it be maintained?

@andy31415 yes, I plan to finalise this and 1 other PR related to time-sync which has been hanging for a long time now.

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.h:
```diff
@@ -104,13 +89,28 @@ class TimeSynchronizationServer : public FabricTable::Delegate
+     * If the DST offset has changed from the previous state or value, the function will emit a DSTStatus event.
+     *
+     * Returns kInvalid if an error occurrs or the DSTOffset list is empty.
+     * Returns kActive if a DST offset is being used.
```
Can this offset that's being used be 0?  Would be good to make that clear.

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.h:
```diff
@@ -104,13 +89,28 @@ class TimeSynchronizationServer : public FabricTable::Delegate
+     *
+     * Returns kInvalid if an error occurrs or the DSTOffset list is empty.
+     * Returns kActive if a DST offset is being used.
+     * Returns kStopped if no DST offset is applicable at the current time.
```
I would have expected kStopped (i.e. "permanent item") to correspond to empty list, but maybe I am missing something? Spec certainly seems to allow empty list, right?

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -555,15 +547,18 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
-        if (tzStore.timeZone.name.HasValue())
+        const TimeSyncDataProvider::TimeZoneStore & tzStore = GetTimeZone()[0];
+        lastTz.offset                                       = tzStore.timeZone.offset;
+        chip::MutableCharSpan tempSpan(name, sizeof(name));
```
```suggestion
        MutableCharSpan tempSpan(name);
```

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -555,15 +547,18 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
         {
-            lastTz.name.SetValue(CharSpan(name));
-            memcpy(name, tzStore.name, sizeof(tzStore.name));
+            lastTz.name.SetValue(CharSpan(tempSpan.data(), tempSpan.size()));
```
```suggestion
            lastTz.name.SetValue(tempSpan);
```

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -616,30 +606,42 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
         if (emit)
         {
-            mEventFlag = TimeSyncEventFlag::kTimeZoneStatus;
+            emitTimeZoneStatusEvent(GetDelegate()->GetEndpoint());
```
Pre-existing, but the right terminology is "generate", not "emit".  Followup to fix that, please?

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -616,30 +606,42 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        // do not propagate if an error occurs here because it is not the user relevant error
```
```suggestion
        // Restore our state to the last known-good one.
        //
        // Do not propagate failure from LoadTimeZone if an error occurs here because it is not the user relevant error
```

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -675,8 +678,10 @@ CHIP_ERROR TimeSynchronizationServer::SetDSTOffset(const DataModel::DecodableLis
-    auto newDstL = dstL.begin();
-    size_t i     = 0;
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    uint64_t lastValidUntil = 0;
```
The use of "goto exit" is pretty unfortunate.  It might have been better to just have two functions: one which returns on failures, and one which calls the other and rolls back the state if failure is returned...  But OK.

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -686,42 +691,44 @@ CHIP_ERROR TimeSynchronizationServer::SetDSTOffset(const DataModel::DecodableLis
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        // do not propagate if an error occurs here because it is not the user relevant error
```
Same suggestion here as for the timezone case.

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -779,13 +787,8 @@ CHIP_ERROR TimeSynchronizationServer::SetUTCTime(EndpointId ep, uint64_t utcTime
-        ChipLogError(Zcl, "Writing TimeSource failed.");
-        return CHIP_IM_GLOBAL_STATUS(Failure);
-    }
+    mGranularity = static_cast<GranularityEnum>(to_underlying(granularity) - 1);
```
I guess the other PR's strict "greater than" check is what keeps this from running off the bottom of the list?

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -381,12 +381,8 @@ void TimeSynchronizationServer::OnTimeSyncCompletionFn(TimeSourceEnum timeSource
-        ChipLogError(Zcl, "Writing TimeSource failed.");
-    }
+    mGranularity = granularity;
+    mTimeSource  = timeSource;
```
Changes to the value of the attribute need to mark it dirty, no?  This applies to a bunch of the attributes in this cluster, actually: it seems to be completely broken wrt subscriptions....

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -893,46 +900,69 @@ TimeState TimeSynchronizationServer::UpdateDSTOffsetState()
+        if (activeDstIndex + 1 >= mDstOffsetObj.validSize)
         {
-            VerifyOrReturnValue(ClearDSTOffset() == CHIP_NO_ERROR, TimeState::kInvalid);
+            // this is the last item in the list that has already expired
```
```suggestion
            // This is the last item in the list, and it has already expired.
```

bzbarsky-apple commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -1202,12 +1232,10 @@ bool emberAfTimeSynchronizationClusterSetTimeZoneCallback(
     if (response.DSTOffsetRequired)
     {
-        TimeState dstState = TimeSynchronizationServer::Instance().UpdateDSTOffsetState();
+        // attempt to emit DSTStatus event if DST was previously available
```
"generate" throughout, not "emit".

fessehaeve committed:
regenerate all after rebasing to latest master

fessehaeve committed:
Update src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

fessehaeve committed:
Update src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

fessehaeve committed:
Update src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

fessehaeve committed:
Update src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

fessehaeve committed:
add a setter for TimeSource and generate report when the value changes
